### PR TITLE
add sap hana support for sql chains and agents

### PR DESCRIPTION
**Pull request by willemi069808**
Added support for SAP HANA databases for SqlDatabaseChain and SqlToolkit.
Main elements:
- added 'sap' as supported database (= typeorm database type name)
- correct SQL generation to use schema name
- added instruction to only query using schema name in prompts
- added 2 examples

This requires `@sap/hana-client` and `hdb-pool` dependencies which I added to the `examples` package.json but not the main package.json.

Database-specific coding seems to have been there for SqlDatabaseChain but not for SqlToolkit. See also #2208

### CODE REVIEW

File: examples/package.json [hunk 1]
```diff
@@ -41,6 +42,7 @@
     "chromadb": "^1.4.0",
     "faiss-node": "^0.2.1",
     "graphql": "^16.6.0",
+    "hdb-pool": "^0.1.6",
```
Comment (jacoblee93) [inline examples/package.json:45]: Where are these required?

File: langchain/src/agents/toolkits/sql/sql.ts [hunk 2]
```diff
@@ -44,7 +45,7 @@ export function createSqlAgent(
   args?: SqlCreatePromptArgs
 ) {
   const {
-    prefix = SQL_PREFIX,
+    prefix = toolkit.db.appDataSourceOptions?.type === 'sap' ? SQL_PREFIX_SAPHANA : SQL_PREFIX,
```
Comment (jacoblee93) [inline langchain/src/agents/toolkits/sql/sql.ts:48]: Let's do this in a separate PR and fix it for all the different SQL dialects?

File: langchain/src/chains/sql_db/sql_db_prompt.ts [hunk 3]
```diff
@@ -102,3 +102,24 @@ Only use the following tables:
 Question: {input}`,
   inputVariables: ["dialect", "table_info", "input", "top_k"],
 });
+
+export const SQL_SAPHANA_PROMPT = /*#__PURE__*/ new PromptTemplate({
```
Comment (jacoblee93) [inline langchain/src/chains/sql_db/sql_db_prompt.ts:106]: Prefer: `SQL_SAP_HANA_PROMPT`

**Comment by jacoblee93** [comment]
Thanks for this! Looks good overall - one question!

File: examples/package.json (hunk 1 above)
Comment (willemi069808) [reply examples/package.json:45]: Good catch. Yes, hdb-pool is implicitly required by `typeorm` even though the package has been deprecated for a while ([hdb-pool github](https://github.com/ckyycc/hdb-pool)) since the features are now supported by NodeJs natively.

`typeorm` allows you to specify your own driver in the connection settings (https://github.com/typeorm/typeorm/blob/master/src/driver/sap/SapDriver.ts#L818), but doing so breaks the code as it is written for hdb-pool only, e.g. expecting an `eventEmitter` property in the driver: (https://github.com/typeorm/typeorm/blob/master/src/driver/sap/SapDriver.ts#L287).

So the `hdb-pool` dependency is needed until `typeorm` updates their code.

File: langchain/src/agents/toolkits/sql/sql.ts (hunk 2 above)
Comment (willemi069808) [reply langchain/src/agents/toolkits/sql/sql.ts:48]: ok - will remove this

File: langchain/src/chains/sql_db/sql_db_prompt.ts (hunk 3 above)
Comment (willemi069808) [reply langchain/src/chains/sql_db/sql_db_prompt.ts:106]: ok - will update

**Comment by willemi069808** [comment]
Please see changes in new commit. Thanks!

File: examples/package.json (hunk 1 above)
Comment (jacoblee93) [reply examples/package.json:45]: Got it - since it's not explicitly in the code, let's not have this in the package.json here and add it to the docs instead.

**Comment by jacoblee93** [comment]
Looks good, one more comment!

**Comment by willemi069808** [comment]
Ok, dependencies removed.

**Comment by jacoblee93** [comment]
Just lint + format + expose an example - thank you!

If you have time to get some of those other agent changes in, we'd welcome it as well :)
